### PR TITLE
Fix `bindname` errors and allow comparing against NULL values

### LINKED ISSUE #5: AttributeError: 'SQLACondition' object has no attribute 'bindname'
Thanks for the great utility!

I frequently run into the error below when the code in `mixins/sqla.py` tries to run a `.ilike( ... )`. All the other operators work fine. Here is a full trace:

```
  File "/opt/bitnami/python/lib/python3.7/site-packages/boolean_parser/parsers/sqla.py", line 54, in filter
    for condition in self.conditions]
  File "/opt/bitnami/python/lib/python3.7/site-packages/boolean_parser/parsers/sqla.py", line 54, in <listcomp>
    for condition in self.conditions]
  File "/opt/bitnami/python/lib/python3.7/site-packages/boolean_parser/parsers/sqla.py", line 54, in filter
    for condition in self.conditions]
  File "/opt/bitnami/python/lib/python3.7/site-packages/boolean_parser/parsers/sqla.py", line 54, in <listcomp>
    for condition in self.conditions]
  File "/opt/bitnami/python/lib/python3.7/site-packages/boolean_parser/mixins/sqla.py", line 135, in filter
    condition = self._filter_one(model, field=field, condition=condition)
  File "/opt/bitnami/python/lib/python3.7/site-packages/boolean_parser/mixins/sqla.py", line 186, in _filter_one
    '%' + bindparam(self.bindname, value) + '%')
AttributeError: 'SQLACondition' object has no attribute 'bindname'
```

### PR DESCRIPTION
**Pull request by linozen**
This fixes #5 and adds the ability to compare against `NULL` in the queries which is something that we need over here: https://git.fsfe.org/fsfe-system-hackers/fsfe-cd/pulls/163.

Many thanks for this great parser. I'm looking forward to including a version of it in the aforementioned project.

### CODE REVIEW

File: boolean_parser/mixins/sqla.py [hunk 1]
```diff
@@ -177,13 +233,31 @@ def _filter_one(self, model, field=None, condition=None):
...
+                if value.lower() == 'null':
+                    condition = field.is_(None)
+                elif value.find('*') >= 0:
                     value = value.replace('*', '%')
-                    condition = field.ilike(
-                        bindparam(self.bindname, value))
+                    condition = field.ilike(value)
+                else:
+                    condition = field.ilike('%' + value + '%')
+            elif isinstance(field.type, sqltypes.BOOLEAN) or \
```
Comment (havok2063) [inline boolean_parser/mixins/sqla.py]: I use `bindparam` to map the value to the full "table.parameter" name.  This makes the internal sql statement syntax a bit more readable, especially when one queries on the same column name from different tables.  This is what `lower_value` is from the output of `bind_and_lower_value` at line 147 does.  

I realized this bit of code never got properly translated and the error you were seeing with `self.bindname` was from old code that no longer exists.  If possible, I'd like to keep the `bindparam` functionality for these cases.  You could use the `self.fullname` instead, and use something like `condition = lower_field.ilike(bindparam(self.fullname, value))` at line 240, and similarly for line 242.

File: boolean_parser/mixins/sqla.py [hunk 2]
```diff
@@ -166,7 +194,35 @@ def _filter_one(self, model, field=None, condition=None):
...
+                isinstance(field.type, sqltypes.Date) or \
+                isinstance(field.type, sqltypes.DateTime):
+                field = getattr(model, self.name)
+                value = self.value
+                if value.lower() == 'null':
+                    condition = field.isnot(None)
+                else:
+                    condition = lower_field.__ne__(lower_value)
+            else:
+                condition = lower_field.__ne__(lower_value)
```
Comment (havok2063) [inline boolean_parser/mixins/sqla.py]: This bit of code looks the same as how it's used for the `==` operator at lines 159-187, with the exception of the `field` function.  I'm always in favor of not duplicating code for easier maintainability and readability.  Maybe we could move this code into a new common method that accepts the relevant inputs?

File: boolean_parser/mixins/sqla.py [hunk 3]
```diff
@@ -166,7 +194,35 @@ def _filter_one(self, model, field=None, condition=None):
...
         elif self.operator == '=':
```
Comment (havok2063) [inline boolean_parser/mixins/sqla.py]: The code under this operator `=` block is also now very similar to the code for the `==` and `!=` operators.  I'm wondering if there's a way to generalize this along with those code blocks and simplify it a bit?  The `ilike` makes it a bit different but just a thought.

**Review by havok2063 — CHANGES_REQUESTED** [review]
This looks good!  I added a few comments in places.  Can you also make sure all the existing tests pass, and add new tests in `tests/parsers/test_sqla.py` for all your new use cases.  It looks like you've added the NULL option, boolean option, and added datetime string formats. And you've expanded the `==` and `!=` operator cases.  I'd like to make sure we capture those in tests.  Later I can use those test examples to update the documentation.

File: boolean_parser/mixins/sqla.py (hunk 1 above)
Comment (havok2063) [reply boolean_parser/mixins/sqla.py]: And similarly in the other code blocks where `ilike` is used.

Comment (linozen) [reply boolean_parser/mixins/sqla.py]: I've added the `bindparam` functionality. Thanks for the pointers.

File: boolean_parser/mixins/sqla.py (hunk 2 above)
Comment (linozen) [reply boolean_parser/mixins/sqla.py]: My lack of understanding of the code led to quite a lot of duplication. It should be much more straightforward now.

File: boolean_parser/mixins/sqla.py (hunk 3 above)
Comment (linozen) [reply boolean_parser/mixins/sqla.py]: see above.

**Comment by linozen** [comment]
Hi @havok2063, many thanks for your review. I've now added a couple more tests and enhanced the test database with dates and nulls. Let me know what you think. Looking forward to getting this merged.

**Review by havok2063 — APPROVED** [review]

**Comment by havok2063** [comment]
@linozen These changes look good to me.  Thanks a lot!

**Comment by linozen** [comment]
Nice! thank you, @havok2063 for merging. Is there any chance you could publish a new release of this package that includes this PR?

**Comment by havok2063** [comment]
@linozen Done.  0.1.3. just released.

**Pull request merged**